Add Open Energy Modeling at JuMP-dev 2024 (#156)
jump-dev/jump-dev.github.io: _posts/2024-09-17-open-energy-models.md

---
layout: post
title:  "Open Energy Modeling at JuMP-dev"
date:   2024-09-17
categories: [open-energy-models]
author: "Oscar Dowson"
---

> In July 2024, we held [JuMP-dev 2024](/meetings/jumpdev2024/), the 7th edition

Author: ```suggestion
In July 2024, we held [JuMP-dev 2024](/meetings/jumpdev2024/), the seventh edition
```

   Pyomo has long supported physical units, and SINTEF have developed [UnitJuMP.jl](https://github.com/trulsf/UnitJuMP.jl).
   Adding support for physical units is an [open issue in JuMP (#1350)](https://github.com/jump-dev/JuMP.jl/issues/1350)
   and is on our [roadmap](https://jump.dev/JuMP.jl/stable/developers/roadmap/).
 * We need better tools for debugging JuMP models. Despite making it easy for
   users to formulate and solve a wide range of optimization problems, JuMP
   provides little support for the users who make mistakes, or tools for
   advanced users to debug problematic models. Moreover, in our experience the
   majority of (expensive) human programmer time is spent, not on formulating or
   solving a model, but on the debugging and testing stage of development. There
   is an [open issue in JuMP (#3664)](https://github.com/jump-dev/JuMP.jl/issues/3664)
   that contains some preliminary ideas for features we could add that would
   help users debug and test JuMP models.

To summarize, JuMP is a powerful tool for energy system modeling. The insights
from JuMP-dev 2024 will help guide future developments, particularly in areas

> like debugging, physical units, and sparse variabble support.

Reviewer: ```suggestion
like debugging, physical units, and sparse variable support.
```

 13. [[2021] AnyMOD.jl: A Julia package for creating energy system models](#2021-anymodjl-a-julia-package-for-creating-energy-system-models)
 14. [[2021] Power Market Tool (POMATO)](#2021-power-market-tool-pomato)
 15. [[2021] UnitCommitment.jl Security-Constrained Unit Commitment in JuMP](#2021-unitcommitmentjl-security-constrained-unit-commitment-in-jump)
 16. [[2021] A Brief Introduction to InfrastructureModels](#2021-a-brief-introduction-to-infrastructuremodels)
 17. [[2019] PowerSimulations.jl](#2019-powersimulationsjl)
 18. [[2017] Stochastic programming in energy systems](#2017-stochastic-programming-in-energy-systems)
 19. [[2017] PowerModels.jl: a Brief Introduction](#2017-powermodelsjl-a-brief-introduction)

## [2024] Applied optimization with JuMP at SINTEF

_Speaker: Truls Flatberg @trulsf_

<iframe width="560" height="315" src="https://www.youtube.com/embed/-a9-ToFiT8E?si=rpQeHt1__-lwxjXT" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

> In this [prize winning](prize/jump-dev-2024) talk, Truls discussed how they have

Reviewer: ```suggestion
In this [prize-winning](prize/jump-dev-2024) talk, Truls discussed how they have
```

a mix of AMPL and GAMS, but they are transitioning to JuMP for new projects. He
mentioned that a key benefit for them is the modularity of JuMP, which is in
contrast to monolithic Mosel files with _O(10⁴)_ lines of code.

One line from his talk stuck out to me: "I think [...] the step from academic
problems to more industrial problems is not the mathematics, but that they are
large scale, so [...] they have to be correct and they have to be robust." The
JuMP developers have recently been discussing how, despite making it easy for
users to formulate and solve a wide range of optimization problems, JuMP
provides little support for the users who make mistakes, or tools for advanced
users to debug problematic models. Moreover, in our experience the majority of
(expensive) human programmer time is spent, not in formulating or solving a
model, but in the debugging and testing stage of development. I have opened the
issue [JuMP#3664](https://github.com/jump-dev/JuMP.jl/issues/3664) with some
preliminary ideas for tools that we could develop to test and debug JuMP models,
and I think these would be highly useful to users.

> Truls gave a shout out to a tutorial I wrote, [Design patters for large models](https://jump.dev/JuMP.jl/stable/tutorials/getting_started/design_patterns_for_larger_models/),

Reviewer: ```suggestion
Truls gave a shout out to a tutorial I wrote, [Design patterns for large models](https://jump.dev/JuMP.jl/stable/tutorials/getting_started/design_patterns_for_larger_models/),
```

SpineOpt.jl repository. If I had to summarize the common root cause of the
issues, it is that managing the input and output of a large quantity of JuMP
models is hard! In Spine's case, a frequent use-case is solving a year of power
system operational problems at hourly resolution via a rolling horizon model of
48 hour blocks rolling forward 24 hours at each step.

Diego's experience reinforces for me the need to include rolling horizon
simulations in our set of benchmark instances that we are building in the
[jump-dev/open-energy-modeling-benchmarks](https://github.com/jump-dev/open-energy-modeling-benchmarks)
repository.

[_Back to contents_](#contents)

## [2024] Solving the Market-to-Market Problem in Large Scale Power Systems

_Speaker: Jose Daniel Lara @jd-lara_

<iframe width="560" height="315" src="https://www.youtube.com/embed/N-jDHickaTc?si=T7eQnTopbCCtES3X" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk Jose Daniel gave an overview and spoke about recent updates to

> [Sienna](https://www.nrel.gov/analysis/sienna.html) (nee PowerSimulations.jl).

Reviewer: ```suggestion
[Sienna](https://www.nrel.gov/analysis/sienna.html) (originally named SIIP).
```

The talk follows his previous talks at [JuMP-dev 2019](#2019-powersimulationsjl)
and [JuMP-dev 2023](https://www.youtube.com/watch?v=J7VbCKsnTvQ).

> A key feature of Sienna is that it is designed for problems with _O(10⁸)_

Reviewer: ```suggestion
A key feature of the Sienna set of tools is that it is designed for problems with _O(10⁸)_
```

> variables and constraints. They are building and solving simulation models of

Reviewer: ```suggestion
A key feature of the Sienna set of tools is that it is designed for problems with _O(10⁸)_
```

> the Eastern Interconnection, which is one of, it not the, largest power system

Reviewer: ```suggestion
the Eastern Interconnection, which is one of, if not the, largest power system
```

ParametricOptInterface.jl in our [jump-dev/open-energy-modeling-benchmarks](https://github.com/jump-dev/open-energy-modeling-benchmarks)
repository, and that JuMP's ability to efficiently solve a sequence of similar
problems with some changes in the problem parameters is both a main selling
point of using JuMP over alternative tools and a current bottleneck in existing
workflows.

[_Back to contents_](#contents)

## [2024] PiecewiseAffineApprox.jl

_Speaker: Lars Hellemo @hellemo_

<iframe width="560" height="315" src="https://www.youtube.com/embed/rm292x59Yjk?si=8lN8MDgFQgUvsHZq" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk, Lars presented his work on a new package,
[PiecewiseAffineApprox.jl](https://github.com/sintefore/PiecewiseAffineApprox.jl),
which automatically approximates nonlinear functions with a linearized MIP
formulation that can be used in a linear solver like HiGHS. The package is used
by SINTEF in many of their energy-related applications.

> JuMP core developer @joehuchette wrote a similar package,

Reviewer: ```suggestion
JuMP core developer Joey Hutchette @joehuchette wrote a similar package,
```

broad range of functionality for constructing piecewise linear approximations.

[_Back to contents_](#contents)

## [2023] How JuMP Enables Abstract Energy System Models

_Speaker: Stefan Strömer @sstroemer_

<iframe width="560" height="315" src="https://www.youtube.com/embed/IFI-u6TiuBk?si=Jco2_udi2BcuDNuj" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk, Stefan discussed an energy system model they have been developing.
A unique feature of their model is that they want to scale from low-latency
model predictive control to continental scale models. Their approach is no-code,
with the complete input described in YAML and CSV files.

Stefan again gives an example that their initial experience with JuMP was not
good because it did not fit their object-orientated design philosophy. However,
after working with it for a while and changing their approach, they are now very
happy with it and with its performance.

Stefan also talked about the need for parameters, and how `parameter * variable`

> support would be useful, but that it complicates. We had

Reviewer: ```suggestion
support would be useful, but that it complicates matters. We had
```

[discussions](https://discourse.julialang.org/t/moi-re-evaluate-custom-bridge-direct-model-bridging/94219)
on the community forum, and Stefan opened a [PR with over 50 comments](https://github.com/jump-dev/MathOptInterface.jl/pull/2092)
that was ultimately closed without merging. This is a tricky topic for which we
do not have a good solution.

[_Back to contents_](#contents)

## [2023] TimeStruct.jl: Multi Horizon Time Modelling in JuMP

_Speaker: Truls Flatberg @trulsf_

<iframe width="560" height="315" src="https://www.youtube.com/embed/Hz6AL5kClKU?si=X2oqhPM3OaqaNsAe" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk Truls presented his work on [TimeStruct.jl](https://github.com/sintefore/TimeStruct.jl),
(if you watch the video, the package is now registered!) which is a JuMP
extension for working with time-structured models that often arise in planning
problems.

Truls points out that a key feature of JuMP is its modularity, and how they can

> invest in the develop of small utility packages like TimeStruct.jl, and then

Reviewer: ```suggestion
invest in the development of small utility packages like TimeStruct.jl, and then
```

re-use the utility package across many different application models.

If you are building a JuMP model with time as a significant component, we
recommend that you try out
[TimeStruct.jl](https://github.com/sintefore/TimeStruct.jl).

[_Back to contents_](#contents)

## [2023] Designing a Flexible Energy System Model Using Multiple Dispatch

_Speaker: Julian Straus @JulStraus_

<iframe width="560" height="315" src="https://www.youtube.com/embed/fARbeM8sANA?si=tETN7GSZNzW5ZzCU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk, Julian discusses [EnergyModelsX](https://github.com/EnergyModelsX)
that he is building at SINTEF.

I found it useful to compare the talk to Jose Daniel's
[[2019] PowerSimulations.jl](#2019-powersimulationsjl).
If multiple groups arrive at the same set of ideas, I think it demonstrates that
hierarchical models that leverage Julia's multiple dispatch is the "right" way
to build large-scale JuMP models.

This approach is the topic of my tutorial,

> [Design patters for large models](https://jump.dev/JuMP.jl/stable/tutorials/getting_started/design_patterns_for_larger_models/),

Reviewer: ```suggestion
[Design patterns for large models](https://jump.dev/JuMP.jl/stable/tutorials/getting_started/design_patterns_for_larger_models/),
```

## [2022] Benchmarking Nonlinear Optimization with AC Optimal Power Flow

_Speaker: Carleton Coffrin @ccoffrin_
<iframe width="560" height="315" src="https://www.youtube.com/embed/tvBNQcuU-hY?si=fSJKG_OK-RhkXAnq" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

This talk by Carleton isn't really about energy modeling, but it is about
solving nonlinear AC Optimal Power Flow problems. I'm including it here for
reader interest because it demonstrates that JuMP is a great tool for building
and solving energy-related optimization problems.

[_Back to contents_](#contents)

## [2021] Modelling Australia's National Electricity Market with JuMP

_Speaker: James Foster @jd-foster_

<iframe width="560" height="315" src="https://www.youtube.com/embed/gbSVH8Q0xq4?si=IeZgeipH3UhTTnkz" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

> In this talk, James provided a general overview of develop models in JuMP. A

Reviewer: ```suggestion
In this talk, James provided a general overview of developing models in JuMP. A
```

loading the data into SQLite tables, performing the computationally intensive
transformation steps using SQL, and then outputting clean tables to DataFrames
for use in JuMP.

This is the same approach that I have taken in tutorials like
[The network multi-commodity flow problem](https://jump.dev/JuMP.jl/stable/tutorials/linear/multi_commodity_network/),
so I wonder if it would be helpful for other users if we advertised this more
widely.

James also raises the reoccurring theme that scaling up your model will present
different issues than the model design phase.

[_Back to contents_](#contents)

## [2021] AnyMOD.jl: A Julia package for creating energy system models

_Speaker: Leonard Goeke @leonardgoeke_

<iframe width="560" height="315" src="https://www.youtube.com/embed/QE_tNDER0F4?si=8TTH54xkuz88tnmY" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

In this talk, Leonard presented his work on [AnyMOD.jl](https://github.com/leonardgoeke/AnyMOD.jl).

> Like Stefan, AnyMOD.jl is a low-code model, in which all inputs and outputs are

Reviewer: ```suggestion
Like Stefan's approach, AnyMOD.jl is a low-code model, in which all inputs and outputs are
```

This is a general interest talk by Carleton isn't really about energy modeling,
but it is about the much larger suite of models around
[InfrastructureModels.jl](https://github.com/lanl-ansi/InfrastructureModels.jl)
that they are building at LANL.

If you are writing energy system models, InfrastructureModels.jl and related
packages like PowerModels.jl are a good place to look for inspiration.

[_Back to contents_](#contents)

## [2019] PowerSimulations.jl

_Speaker: Jose Daniel Lara @jd-lara_

<iframe width="560" height="315" src="https://www.youtube.com/embed/JAHjZYiIJeI?si=euyA8WzK-lEs_Pg9" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

This talk is the first talk by Jose Daniel at a JuMP-dev on PowerSimulations.jl

> (now called Sienna). I found it useful to compare this talk to

Reviewer: ```suggestion
(now part of Sienna). I found it useful to compare this talk to
```

hits upon on many of the same themes that Julian did in his 2023 talk
[[2023] Designing a Flexible Energy System Model Using Multiple Dispatch](#2023-designing-a-flexible-energy-system-model-using-multiple-dispatch).
This suggests to me that we can better share lessons learned between developers.

Jose Daniel also talks about the issue with parameters and time-series data.
Five years later and we still do not have a perfect solution to this.

Finally, Jose Daniel talked about how they wanted to scale to 50,000 buses.
Well, in his [2024 talk](#2024-solving-the-market-to-market-problem-in-large-scale-power-systems),
he mentioned that Sienna now runs on problems with 150,000 buses. It's nice to
see progress!

[_Back to contents_](#contents)

## [2017] Stochastic programming in energy systems

_Speaker: Joaquim Dias Garcia @joaquimg_

<iframe width="560" height="315" src="https://www.youtube.com/embed/HwOOww8vwyA?si=Hs065CFXrIe2kg6q" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

> In this talk from thhe first JuMP-dev, Joaquim discussed how PSR were using JuMP

Author: ```suggestion
In this talk from the first JuMP-dev, Joaquim discussed how PSR were using JuMP
```

to solve a variety of energy-related models, both for research and for
industrial clients around the world.

For me, it was interesting to revisit this talk because Joaquim's discussion
about stochastic dual dynamic programming (I develop [SDDP.jl](https://sddp.dev).)

> Joaquim described how they implemmented some of the algorithm in JuMP, and some

Reviewer: ```suggestion
Joaquim described how they implemented some of the algorithm in JuMP, and some
```
